### DISCUSSION TITLE
Windows - Specifying client certificate using EKU fails [connection error]

### DISCUSSION THREAD

**Discussion by acrolink** [strongswan/strongswan#569] [Q&A]
I connect successfully to the VPN server using client certificate (imported into Windows certificate manager) and public key in Windows 10 using this:

![image](https://user-images.githubusercontent.com/13219117/131266345-c01a45ed-4b75-41d5-97a1-7fa285fcd970.png)

However, if I instruct Windows to use a particular client certificate when connecting (using `EKU` and `OID`), the connection fails. I try to do that like this:

![image](https://user-images.githubusercontent.com/13219117/131266349-d2961ba6-3977-4277-9aef-2f6b892accf1.png)

![image](https://user-images.githubusercontent.com/13219117/131266116-4d89b135-3a37-4be6-9aeb-3554a6996e27.png)

![image](https://user-images.githubusercontent.com/13219117/131266132-1f587ec1-9a84-4379-a41b-8f6ec78193b5.png)

If I do this, I get:

![image](https://user-images.githubusercontent.com/13219117/131266195-5f56ee02-98d7-457c-93d5-5a379d178e89.png)

Here is my `ipsec.conf`:

```
conn win7
     leftauth=pubkey
     leftid=@myserver.com
     leftsendcert=always
     leftsubnet=0.0.0.0/0
     rightsourceip=10.99.10.10/32
     rightid="C=IL, O=???, CN=???@???.???"
     keyexchange=ikev2
     authby=pubkey
     auto=add
```

What can be the problem here? Thank you.

**Comment by Thermi** [comment]
> However, if I instruct Windows to use a particular client certificate when connecting (using EKU and OID), the connection fails. I try to do that like this:

That's extremely vague.
Please follow the instructions in #196, as linked in the red-orange banner of the discussion page.

**Comment by tobiasbrunner** [comment]
`Use machine certificates` uses plain IKEv2 certificate authentication. But you then changed that to EAP, which, with certificates, results EAP-TLS authentication (user certificates, which are also stored in a different trust store). So you have to modify the server config, see [WindowsClients](https://wiki.strongswan.org/projects/strongswan/wiki/WindowsClients#B-Authentication-using-X509-User-Certificates).

**Reply by acrolink** [reply]
> `Use machine certificates` uses plain IKEv2 certificate authentication. But you then changed that to EAP, which, with certificates, results EAP-TLS authentication (user certificates, which are also stored in a different trust store). So you have to modify the server config, see [WindowsClients](https://wiki.strongswan.org/projects/strongswan/wiki/WindowsClients#B-Authentication-using-X509-User-Certificates).

Exactly, I have to configure the server to use `EAP-TLS` or `EAP-TTLS`. I tried to do that and I have installed the `charon` plugin providing `EAP-TLS` but still the server doesn't offer that mechanism to the Windows client.

Basically, How to enable `EAP-TLS` on the server ? Is it done within `ipsec.conf` only or are there any additional files to be modified ?

Here is the log:

```

Aug 30 12:35:26 ac-km-01 charon[35334]: 16[ENC] splitting IKE message (1516 bytes) into 2 fragments
Aug 30 12:35:26 ac-km-01 charon[35334]: 16[ENC] generating IKE_AUTH response 1 [ EF(1/2) ]
Aug 30 12:35:26 ac-km-01 charon[35334]: 16[ENC] generating IKE_AUTH response 1 [ EF(2/2) ]
Aug 30 12:35:26 ac-km-01 charon[35334]: 16[NET] sending packet: from server-ip[4500] to x.x.x.x[4500] (1248 bytes)
Aug 30 12:35:26 ac-km-01 charon[35334]: 16[NET] sending packet: from server-ip[4500] to x.x.x.x[4500] (336 bytes)
Aug 30 12:35:26 ac-km-01 charon[35334]: 05[NET] received packet: from x.x.x.x[4500] to server-ip[4500] (76 bytes)
Aug 30 12:35:26 ac-km-01 charon[35334]: 05[ENC] parsed IKE_AUTH request 2 [ EAP/RES/NAK ]
Aug 30 12:35:26 ac-km-01 charon[35334]: 05[IKE] received EAP_NAK, sending EAP_FAILURE
Aug 30 12:35:26 ac-km-01 charon[35334]: 05[ENC] generating IKE_AUTH response 2 [ EAP/FAIL ]

```

**Reply by tobiasbrunner** [reply]
> Basically, How to enable `EAP-TLS` on the server ? Is it done within `ipsec.conf` only or are there any additional files to be modified ?

`rightauth=eap-tls`, refer to the docs for details (in particular the page I linked above).

> Here is the log:

It's not, it's a snippet and we clearly don't see what EAP method was actually requested.
